File: ts-sdk/src/types/icon.ts
/**
 * `Icon` and `Icons` types with security validation (§14.2).
 *
 * `Icon` describes a single renderable icon image. `Icons` is the mixin that
 * contributes an OPTIONAL `icons` array to identity and descriptor objects.
 *
 * Security model (§14.2):
 *  - Only `https:` URLs and `data:` URIs are accepted (R-14.2-o).
 *  - `javascript:`, `file:`, `ftp:`, `ws:` and other unsafe schemes are rejected (R-14.2-n).
 *  - MIME type is detected from magic bytes, not from the declared type (R-14.2-s).
 *  - Only image types on the allowlist are rendered (R-14.2-u).
 *  - Minimum required formats: image/png and image/jpeg (R-14.2-l).
 *  - Recommended additional formats: image/svg+xml and image/webp (R-14.2-m).
 */

import { z } from 'zod';

/** Background theme the icon is designed for (§14.2). */
export type IconTheme = 'light' | 'dark';

/**
 * `Icon` schema (§14.2, R-14.2-c – R-14.2-j).
 *
 * `src` is REQUIRED; all other fields are OPTIONAL.
 * The `theme` field is a closed enum: only `"light"` and `"dark"` are valid. (R-14-a)
 * Additional unknown fields are ignored via `.passthrough()` (§2.3.4).
 */
export const IconSchema = z.object({
  /** REQUIRED. URI pointing to the icon resource (https: URL or data: URI). (R-14.2-c) */
  src: z.string(),
  /** OPTIONAL. MIME-type override when the source type is missing or generic. (R-14.2-g) */
  mimeType: z.string().optional(),
  /**
   * OPTIONAL. Intended-use sizes; each entry is `"WxH"` or `"any"` for scalable. (R-14.2-h)
   * Omitted ⇒ usable at any size. (R-14.2-i)
   */
  sizes: z.array(z.string().regex(/^\d+x\d+$|^any$/)).optional(),
  /**
   * OPTIONAL. Background theme the icon is designed for. (R-14.2-j)
   * Omitted ⇒ usable with any theme. (R-14.2-k)
   */
  theme: z.enum(['light', 'dark']).optional(),
}).passthrough();

export type Icon = z.infer<typeof IconSchema>;

/** `Icons` mixin schema — contributes the OPTIONAL `icons` array. (R-14.2-b, R-14.2-v) */
export const IconsSchema = z.object({
  /** OPTIONAL. A set of sized icons a consumer MAY display. Absent ⇒ no icons advertised. */
  icons: z.array(IconSchema).optional(),
});

export type Icons = z.infer<typeof IconsSchema>;

/** MIME types a consumer MUST support when rendering icons. (R-14.2-l, AC-20.19) */
export const REQUIRED_IMAGE_TYPES = new Set(['image/png', 'image/jpeg', 'image/jpg']);

/** MIME types a consumer SHOULD additionally support. (R-14.2-m, AC-20.20) */
export const RECOMMENDED_IMAGE_TYPES = new Set(['image/svg+xml', 'image/webp']);

/** Default allowlist: REQUIRED + RECOMMENDED types. (R-14.2-u) */
export const DEFAULT_IMAGE_ALLOWLIST = new Set([
  ...REQUIRED_IMAGE_TYPES,
  ...RECOMMENDED_IMAGE_TYPES,
]);

/** Error thrown when an icon URI or its content is rejected for security reasons. */
export class IconValidationError extends Error {
  constructor(
    public readonly src: string,
    reason: string,
  ) {
    super(`Icon rejected (${reason}): ${src}`);
    this.name = 'IconValidationError';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/**
 * Validates an icon `src` URI against the security rules (§14.2).
 *
 * A consumer MUST accept only `https:` URLs or `data:` URIs. (R-14.2-o, AC-20.22)
 * A consumer MUST reject `javascript:`, `file:`, `ftp:`, `ws:`, and other
 * unsafe schemes. (R-14.2-n, AC-20.21)
 *
 * Note: `http:` is also rejected because R-14.2-o's stricter rule governs
 * consumer acceptance and supersedes the field description in R-14.2-d.
 *
 * @throws {IconValidationError} When the scheme is not `https:` or `data:`.
 */
export function validateIconSrc(src: string): void {
  const colon = src.indexOf(':');
  if (colon === -1) {
    throw new IconValidationError(src, 'no URI scheme present');
  }
  const scheme = src.slice(0, colon + 1).toLowerCase();
  if (scheme !== 'https:' && scheme !== 'data:') {
    throw new IconValidationError(
      src,
      `scheme '${scheme}' is not permitted; only https: and data: are accepted`,
    );
  }
}

/** Returns `true` when `src` passes `validateIconSrc` without throwing. (AC-20.21, AC-20.22) */
export function isValidIconSrc(src: string): boolean {
  try {
    validateIconSrc(src);
    return true;
  } catch {
    return false;
  }
}

/** Magic byte signatures for supported image types. (R-14.2-s, AC-20.26) */
const MAGIC_BYTES: ReadonlyArray<{ mimeType: string; signature: readonly number[] }> = [
  { mimeType: 'image/png', signature: [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a] },
  { mimeType: 'image/jpeg', signature: [0xff, 0xd8, 0xff] },
  { mimeType: 'image/gif', signature: [0x47, 0x49, 0x46] },
  { mimeType: 'image/webp', signature: [0x52, 0x49, 0x46, 0x46] }, // RIFF; verify WEBP at offset 8
];

/**
 * Detects the MIME type of an image from its magic bytes, treating the
 * declared MIME type as advisory only. (R-14.2-s, AC-20.26)
 *
 * Returns `null` when no known signature matches.
 */
export function detectMimeTypeFromMagicBytes(bytes: Uint8Array): string | null {
  for (const { mimeType, signature } of MAGIC_BYTES) {
    if (signature.every((b, i) => bytes[i] === b)) {
      if (mimeType === 'image/webp') {
        // RIFF container: bytes 8–11 must be 'WEBP' (0x57 0x45 0x42 0x50)
        const webpTag = [0x57, 0x45, 0x42, 0x50];
        if (!webpTag.every((b, i) => bytes[8 + i] === b)) continue;
      }
      return mimeType;
    }
  }
  // SVG: XML-based, no magic bytes — detect by leading text
  if (bytes.length >= 4) {
    const head = new TextDecoder('utf-8', { fatal: false }).decode(bytes.slice(0, 100));
    const trimmed = head.trimStart().toLowerCase();
    if (trimmed.startsWith('<?xml') || trimmed.startsWith('<svg')) {
      return 'image/svg+xml';
    }
  }
  return null;
}

/**
 * Validates an icon's byte content before rendering. (R-14.2-r – R-14.2-u, AC-20.25–28)
 *
 * 1. Detects the actual MIME type from magic bytes (ignores declared type).
 * 2. Rejects when the detected type is unknown.
 * 3. When `declaredMimeType` is provided, rejects on mismatch.
 * 4. Rejects types outside the `allowedTypes` set.
 *
 * @returns The detected MIME type on success.
 * @throws {IconValidationError} On any validation failure.
 */
export function validateIconBytes(
  bytes: Uint8Array,
  declaredMimeType?: string,
  allowedTypes: ReadonlySet<string> = DEFAULT_IMAGE_ALLOWLIST,
): string {
  const detected = detectMimeTypeFromMagicBytes(bytes);

  if (!detected) {
    throw new IconValidationError('(bytes)', 'unknown image type; cannot render');
  }

  if (!allowedTypes.has(detected)) {
    throw new IconValidationError('(bytes)', `image type ${detected} is not on the allowlist`);
  }

  if (declaredMimeType) {
    // Normalise image/jpg → image/jpeg before comparison
    const norm = (t: string) => (t === 'image/jpg' ? 'image/jpeg' : t);
    if (norm(detected) !== norm(declaredMimeType)) {
      throw new IconValidationError(
        '(bytes)',
        `MIME type mismatch: declared '${declaredMimeType}', detected '${detected}'`,
      );
    }
  }

  return detected;
}

// ─── Secure icon fetch (§14.2) ──────────────────────────────────────────────────

/** Result of {@link fetchIcon}: the validated image bytes, detected type, and final URL. */
export interface FetchIconResult {
  /** The fetched image bytes. */
  bytes: Uint8Array;
  /** The MIME type detected from magic bytes (R-14.2-s). */
  mimeType: string;
  /** The URL the bytes were ultimately read from (same origin as `src`). */
  finalUrl: string;
}

/** Options for {@link fetchIcon}. */
export interface FetchIconOptions {
  /** Override the `fetch` implementation (injection point for tests / non-global runtimes). */
  fetch?: typeof fetch;
  /** Allowed rendered MIME types; defaults to {@link DEFAULT_IMAGE_ALLOWLIST}. */
  allowedTypes?: ReadonlySet<string>;
  /** Maximum number of same-origin redirects to follow before giving up. Default 5. */
  maxRedirects?: number;
}

/** The HTTP status codes that denote a redirect. */
function isRedirectStatus(status: number): boolean {
  return status === 301 || status === 302 || status === 303 || status === 307 || status === 308;
}

/** Decodes a `data:` URI's payload to bytes (base64 or percent-encoded). */
function decodeDataUri(uri: string): Uint8Array {
  const comma = uri.indexOf(',');
  if (comma === -1) throw new IconValidationError(uri, 'malformed data: URI (missing comma)');
  const meta = uri.slice('data:'.length, comma);
  const payload = uri.slice(comma + 1);
  if (/;base64$/i.test(meta)) {
    const binary = atob(payload);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
    return bytes;
  }
  return new TextEncoder().encode(decodeURIComponent(payload));
}

/**
 * Securely fetches and validates an icon. Edge-friendly: built on Web-platform
 * `fetch` only (no `node:*`), so it runs on Node, Cloudflare Workers, Deno, and
 * browsers.
 *
 * Security rules enforced (§14.2):
 *  - `src` MUST be `https:` or `data:` (R-14.2-o, via {@link validateIconSrc}).
 *  - Redirects are followed manually; a redirect that changes the scheme or
 *    moves to a different origin MUST NOT be followed and is rejected
 *    (R-14.2-p, TV-20.12).
 *  - The request is credential-free: `credentials: 'omit'` and no `Authorization`
 *    or `Cookie` header is ever sent (R-14.2-q, TV-20.13).
 *  - The returned bytes are validated against the allowlist by magic bytes,
 *    ignoring the declared type (R-14.2-r – R-14.2-u, via {@link validateIconBytes}).
 *
 * @throws {IconValidationError} On a disallowed scheme, a cross-origin/scheme-change
 *   redirect, a non-2xx status, too many redirects, or invalid image bytes.
 */
export async function fetchIcon(src: string, options: FetchIconOptions = {}): Promise<FetchIconResult> {
  validateIconSrc(src); // R-14.2-o: only https: or data:
  const allowed = options.allowedTypes ?? DEFAULT_IMAGE_ALLOWLIST;

  // `data:` icons carry their bytes inline — no network request, nothing to redirect.
  if (src.toLowerCase().startsWith('data:')) {
    const bytes = decodeDataUri(src);
    return { bytes, mimeType: validateIconBytes(bytes, undefined, allowed), finalUrl: src };
  }

  const fetchImpl = options.fetch ?? globalThis.fetch;
  if (typeof fetchImpl !== 'function') {
    throw new IconValidationError(src, 'no global `fetch` available; pass options.fetch');
  }

  const origin = new URL(src);
  let current = origin;
  const maxRedirects = options.maxRedirects ?? 5;

  for (let hop = 0; hop <= maxRedirects; hop++) {
    // Credential-free request, manual redirect handling. (R-14.2-q, R-14.2-p)
    const response = await fetchImpl(current.toString(), {
      method: 'GET',
      redirect: 'manual',
      credentials: 'omit',
      headers: {},
    });

    if (isRedirectStatus(response.status)) {
      const location = response.headers.get('location');
      if (!location) {
        throw new IconValidationError(src, `redirect ${response.status} without a Location header`);
      }
      const next = new URL(location, current);
      if (next.protocol !== origin.protocol) {
        throw new IconValidationError(
          src,
          `refusing redirect with scheme change '${origin.protocol}' → '${next.protocol}'`,
        );
      }
      if (next.origin !== origin.origin) {
        throw new IconValidationError(
          src,
          `refusing cross-origin redirect '${origin.origin}' → '${next.origin}'`,
        );
      }
      current = next;
      continue;
    }

    if (response.status >= 200 && response.status < 300) {
      const bytes = new Uint8Array(await response.arrayBuffer());
      return { bytes, mimeType: validateIconBytes(bytes, undefined, allowed), finalUrl: current.toString() };
    }

    throw new IconValidationError(src, `icon fetch failed with HTTP ${response.status}`);
  }

  throw new IconValidationError(src, `too many redirects (more than ${maxRedirects})`);
}

File: py-sdk/src/mcp/types/icon.py
"""``Icon`` / ``Icons`` types with security validation (§14.2).

``Icon`` describes a single renderable icon image; ``Icons`` contributes an OPTIONAL
``icons`` array to identity/descriptor objects.

Security model (§14.2): only ``https:`` URLs and ``data:`` URIs are accepted (R-14.2-o);
unsafe schemes are rejected (R-14.2-n); MIME type is detected from magic bytes, not the
declared type (R-14.2-s); only allowlisted image types render (R-14.2-u).

The secure network fetch (:func:`fetch_icon`, the TS SDK's ``fetchIcon``) follows the
same rules: it manually handles redirects, refusing any scheme change or cross-origin
hop (R-14.2-p), and sends a credential-free request (R-14.2-q). The HTTP client is
injectable for testing via the ``fetch`` parameter; the default uses ``httpx`` (the
same dependency :mod:`mcp.client.http` relies on).
"""

from __future__ import annotations

import base64
import binascii
import re
import urllib.parse
from dataclasses import dataclass
from typing import Annotated, Callable, Literal, Protocol, runtime_checkable

from pydantic import AfterValidator

from mcp._model import McpModel, validates

#: Background theme an icon targets (§14.2) — the closed set, as a field type.
IconTheme = Literal["light", "dark"]

#: Background themes an icon may target (§14.2). Closed set.
ICON_THEMES = frozenset({"light", "dark"})

# Each size entry is "WxH" or "any" (for scalable). (R-14.2-h) — fullmatch so a trailing
# newline cannot sneak past, matching JS's `$` (which, unlike Python's, never does).
_SIZES_RE = re.compile(r"\d+x\d+|any")


def _require_icon_size(value: str) -> str:
  """Field validator: reject an icon ``sizes`` entry that is not ``"WxH"`` or ``"any"``."""
  if not _SIZES_RE.fullmatch(value):
    raise ValueError('icon size MUST be "WxH" or "any" (R-14.2-h)')
  return value


#: An icon size string — the analogue of Zod ``z.string().regex(/^\d+x\d+$|^any$/)``.
IconSize = Annotated[str, AfterValidator(_require_icon_size)]


class Icon(McpModel):
  """A single renderable icon image (§14.2) — the Python analogue of the TS ``IconSchema``.

  ``src`` is REQUIRED; all other fields are OPTIONAL. ``theme`` is a closed enum; unknown
  members pass through (forward-compatible).
  """

  #: REQUIRED. URI pointing to the icon resource (https: URL or data: URI). (R-14.2-c)
  src: str
  #: OPTIONAL. MIME-type override when the source type is missing or generic. (R-14.2-g)
  mime_type: str | None = None
  #: OPTIONAL. Intended-use sizes; each entry is ``"WxH"`` or ``"any"``. (R-14.2-h, R-14.2-i)
  sizes: list[IconSize] | None = None
  #: OPTIONAL. Background theme the icon is designed for. (R-14.2-j, R-14.2-k)
  theme: IconTheme | None = None


class Icons(McpModel):
  """The ``Icons`` mixin — contributes the OPTIONAL ``icons`` array. (R-14.2-b, R-14.2-v)"""

  #: OPTIONAL. A set of sized icons a consumer MAY display. Absent ⇒ none advertised.
  icons: list[Icon] | None = None


def is_valid_icon(value: object) -> bool:
  """Return ``True`` for a valid ``Icon`` (§14.2): REQUIRED string ``src``; OPTIONAL
  ``mimeType`` (str), ``sizes`` (list of ``"WxH"``/``"any"``), ``theme`` (light/dark).
  Extra members tolerated.
  """
  return validates(Icon, value)


def is_valid_icons(value: object) -> bool:
  """Return ``True`` for a valid ``Icons`` mixin (§14.2): an object with an OPTIONAL
  ``icons`` array of valid :class:`Icon` entries. An absent or empty array is valid;
  extra members are tolerated. (R-14.2-b, R-14.2-v)
  """
  return validates(Icons, value)


#: MIME types a consumer MUST support when rendering icons. (R-14.2-l)
REQUIRED_IMAGE_TYPES = frozenset({"image/png", "image/jpeg", "image/jpg"})
#: MIME types a consumer SHOULD additionally support. (R-14.2-m)
RECOMMENDED_IMAGE_TYPES = frozenset({"image/svg+xml", "image/webp"})
#: Default allowlist: REQUIRED + RECOMMENDED. (R-14.2-u)
DEFAULT_IMAGE_ALLOWLIST = REQUIRED_IMAGE_TYPES | RECOMMENDED_IMAGE_TYPES


class IconValidationError(Exception):
  """Raised when an icon URI or its content is rejected for security reasons."""

  def __init__(self, src: str, reason: str) -> None:
    super().__init__(f"Icon rejected ({reason}): {src}")
    self.src = src


def validate_icon_src(src: str) -> None:
  """Validate an icon ``src`` URI scheme (§14.2): only ``https:`` or ``data:`` accepted.

  :raises IconValidationError: when the scheme is missing or not permitted. (R-14.2-o,
    R-14.2-n)
  """
  colon = src.find(":")
  if colon == -1:
    raise IconValidationError(src, "no URI scheme present")
  scheme = src[: colon + 1].lower()
  if scheme not in ("https:", "data:"):
    raise IconValidationError(src, f"scheme '{scheme}' is not permitted; only https: and data: are accepted")


def is_valid_icon_src(src: str) -> bool:
  """Return ``True`` when ``src`` passes :func:`validate_icon_src` without raising."""
  try:
    validate_icon_src(src)
    return True
  except IconValidationError:
    return False


#: Magic-byte signatures for supported image types. (R-14.2-s)
MAGIC_BYTES: tuple[tuple[str, bytes], ...] = (
  ("image/png", b"\x89PNG\r\n\x1a\n"),
  ("image/jpeg", b"\xff\xd8\xff"),
  ("image/gif", b"GIF"),
  ("image/webp", b"RIFF"),  # RIFF container; bytes 8-11 must be 'WEBP'
)


def detect_mime_type_from_magic_bytes(data: bytes) -> str | None:
  """Detect an image's MIME type from its magic bytes, treating any declared type as
  advisory (R-14.2-s). Returns ``None`` when no known signature matches.
  """
  for mime_type, signature in MAGIC_BYTES:
    if data[: len(signature)] == signature:
      if mime_type == "image/webp" and data[8:12] != b"WEBP":
        continue
      return mime_type
  # SVG is XML-based (no magic bytes) — detect by leading text.
  if len(data) >= 4:
    head = data[:100].decode("utf-8", errors="ignore").lstrip().lower()
    if head.startswith("<?xml") or head.startswith("<svg"):
      return "image/svg+xml"
  return None


def validate_icon_bytes(
  data: bytes,
  declared_mime_type: str | None = None,
  allowed_types: frozenset[str] | set[str] = DEFAULT_IMAGE_ALLOWLIST,
) -> str:
  """Validate icon byte content before rendering (§14.2). Returns the detected MIME type.

  Detects the actual type from magic bytes (ignoring the declared type), rejects unknown
  or non-allowlisted types, and — when ``declared_mime_type`` is given — rejects a
  mismatch (``image/jpg`` is normalised to ``image/jpeg``). (R-14.2-r–R-14.2-u)

  :raises IconValidationError: on any validation failure.
  """
  detected = detect_mime_type_from_magic_bytes(data)
  if detected is None:
    raise IconValidationError("(bytes)", "unknown image type; cannot render")
  if detected not in allowed_types:
    raise IconValidationError("(bytes)", f"image type {detected} is not on the allowlist")
  if declared_mime_type:
    def _norm(t: str) -> str:
      return "image/jpeg" if t == "image/jpg" else t

    if _norm(detected) != _norm(declared_mime_type):
      raise IconValidationError(
        "(bytes)", f"MIME type mismatch: declared '{declared_mime_type}', detected '{detected}'"
      )
  return detected


# ─── Secure icon fetch (§14.2) ────────────────────────────────────────────────

#: HTTP status codes that denote a redirect (the TS SDK's ``isRedirectStatus``).
_REDIRECT_STATUSES = frozenset({301, 302, 303, 307, 308})


def is_redirect_status(status: int) -> bool:
  """Return ``True`` when ``status`` is an HTTP redirect (301/302/303/307/308)."""
  return status in _REDIRECT_STATUSES


@runtime_checkable
class FetchResponse(Protocol):
  """The minimal HTTP response surface :func:`fetch_icon` consumes.

  Satisfied by ``httpx.Response`` (``status_code``, ``headers``, ``content``); a test
  fetcher need only supply these three members.
  """

  @property
  def status_code(self) -> int: ...

  @property
  def headers(self) -> dict: ...

  @property
  def content(self) -> bytes: ...


#: A ``fetch`` callable: takes a URL, returns a :class:`FetchResponse`. Redirects MUST be
#: surfaced (not auto-followed) so :func:`fetch_icon` can enforce R-14.2-p itself.
Fetch = Callable[[str], FetchResponse]


@dataclass(frozen=True)
class FetchIconResult:
  """The validated outcome of :func:`fetch_icon`."""

  bytes: bytes  #: The fetched image bytes.
  mime_type: str  #: The MIME type detected from magic bytes. (R-14.2-s)
  final_url: str  #: The URL the bytes were ultimately read from (same origin as ``src``).


def _decode_data_uri(uri: str) -> bytes:
  """Decode a ``data:`` URI payload to bytes (Base64 or percent-encoded).

  :raises IconValidationError: when the URI is malformed.
  """
  comma = uri.find(",")
  if comma == -1:
    raise IconValidationError(uri, "malformed data: URI (missing comma)")
  meta = uri[len("data:"):comma]
  payload = uri[comma + 1:]
  if re.search(r";base64$", meta, re.IGNORECASE):
    try:
      return base64.b64decode(payload, validate=True)
    except (binascii.Error, ValueError) as exc:
      raise IconValidationError(uri, f"malformed base64 data: URI payload ({exc})") from exc
  return urllib.parse.unquote_to_bytes(payload)


def _default_fetch(url: str) -> FetchResponse:
  """The default credential-free, no-auto-redirect fetcher, backed by ``httpx``.

  Imported lazily so the synchronous validation core has no import-time network dependency.
  """
  import httpx

  return httpx.get(
    url,
    follow_redirects=False,  # R-14.2-p: redirects are vetted manually below.
    headers={},  # R-14.2-q: no Authorization / Cookie header.
    # httpx sends no cookies unless a cookie jar is supplied — this request carries none.
  )


def fetch_icon(
  src: str,
  *,
  fetch: Fetch | None = None,
  allowed_types: frozenset[str] | set[str] = DEFAULT_IMAGE_ALLOWLIST,
  max_redirects: int = 5,
) -> FetchIconResult:
  """Securely fetch and validate an icon (§14.2), enforcing the consumer security rules.

  * ``src`` MUST be ``https:`` or ``data:`` (R-14.2-o, via :func:`validate_icon_src`).
  * Redirects are followed manually; a redirect that changes the scheme or moves to a
    different origin MUST NOT be followed and is rejected (R-14.2-p, TV-20.12).
  * The request is credential-free — no ``Authorization`` or ``Cookie`` header is sent
    (R-14.2-q, TV-20.13).
  * The returned bytes are validated against the allowlist by magic bytes, ignoring any
    declared type (R-14.2-r–R-14.2-u, via :func:`validate_icon_bytes`).

  ``data:`` icons carry their bytes inline, so no network request is made.

  :raises IconValidationError: on a disallowed scheme, a cross-origin/scheme-change
    redirect, a non-2xx status, too many redirects, or invalid image bytes.
  """
  validate_icon_src(src)  # R-14.2-o: only https: or data:

  # `data:` icons carry their bytes inline — no network request, nothing to redirect.
  if src.lower().startswith("data:"):
    data = _decode_data_uri(src)
    return FetchIconResult(data, validate_icon_bytes(data, None, allowed_types), src)

  fetcher = fetch or _default_fetch
  origin = urllib.parse.urlsplit(src)
  current = src

  for _ in range(max_redirects + 1):
    response = fetcher(current)  # R-14.2-q: credential-free request (see fetcher contract).
    status = response.status_code
    if is_redirect_status(status):
      # The FetchResponse protocol guarantees ``headers`` is a mapping (httpx.Headers and
      # dict both support .get), so no defensive hasattr guard is needed.
      location = response.headers.get("location")
      if not location:
        raise IconValidationError(src, f"redirect {status} without a Location header")
      next_url = urllib.parse.urljoin(current, location)
      nxt = urllib.parse.urlsplit(next_url)
      if nxt.scheme != origin.scheme:
        raise IconValidationError(
          src, f"refusing redirect with scheme change '{origin.scheme}' → '{nxt.scheme}'"
        )
      if (nxt.scheme, nxt.netloc) != (origin.scheme, origin.netloc):
        raise IconValidationError(
          src,
          f"refusing cross-origin redirect '{origin.scheme}://{origin.netloc}' → "
          f"'{nxt.scheme}://{nxt.netloc}'",
        )
      current = next_url
      continue

    if 200 <= status < 300:
      data = bytes(response.content)
      return FetchIconResult(data, validate_icon_bytes(data, None, allowed_types), current)

    raise IconValidationError(src, f"icon fetch failed with HTTP {status}")

  raise IconValidationError(src, f"too many redirects (more than {max_redirects})")
